[2019.2.1] Bring nox into 2019.2.1

s0undt3ch opened the pull request:
### What does this PR do?
See title.
More or less what's done in https://github.com/saltstack/salt/pull/52848 but cherry picking the changes.

s0undt3ch committed:
Ignore `.nox` directory

s0undt3ch committed:
Rerun failed tests

s0undt3ch committed:
IOFLO is pulled in from the raet requirements

s0undt3ch committed:
Previously generated requirements were py2 only

s0undt3ch committed:
Kubernetes 3.0.0 does include the requirements files. pip-compile chokes on that

s0undt3ch committed:
Don't lock the docker requirement. It's not locked on the other platforms

s0undt3ch committed:
Be aware of the new static requirements layout

s0undt3ch committed:
Add pre-commit config to generate static requirements

s0undt3ch committed:
Update/Generate static requirements to new layout

s0undt3ch committed:
Parametrize the transport

s0undt3ch committed:
Add separate crypto sessions

s0undt3ch committed:
Instead of quoting session parameters, proxy nox sessions

s0undt3ch committed:
PyZMQ is pulled in from zeromq.txt

s0undt3ch committed:
The required crypto library is pulled in from zeromq.txt or raet.txt

s0undt3ch committed:
Static requirements are now placed on `py<major>.<minor>` subdirectories

s0undt3ch committed:
We now compile requirements for each of the supported minor version

s0undt3ch committed:
We want a more verbose pip-compile output

s0undt3ch committed:
We don't test OSX nor Windows under Py3.4

s0undt3ch committed:
Update(fix) requirements

s0undt3ch committed:
Use pycryptodomex on windows and PyCrypto on the rest

Because PyCrypto on windows needs to be patched.

s0undt3ch committed:
Remove pycryptodome from compiled requirements files

s0undt3ch committed:
Recompile requirements

s0undt3ch committed:
Remove requiremens for platforms which aren't tested under Py3

s0undt3ch committed:
Bump apache-libcloud requirement because 1.0.0 fails to install on windows

s0undt3ch committed:
Back to `+apache-libcloud==1.0.0` and skip windows for now

s0undt3ch committed:
Lock lint requirements

s0undt3ch committed:
Add nox session/env/target to run lint against Salt and it's test suite

s0undt3ch committed:
Fix lint issues on salt

s0undt3ch committed:
Fix lint issues under tests

s0undt3ch committed:
Lock docs python requirements

s0undt3ch committed:
Add nox env to build docs

s0undt3ch committed:
Don't require six

s0undt3ch committed:
Remove dead code

s0undt3ch committed:
We currently only support running lint under Py2.7

s0undt3ch committed:
Don't repeat filed tests on missing/empty failed test files file

s0undt3ch committed:
Throw error when running either on Py<2.7 or Py>=3.7

s0undt3ch committed:
Compile requirements for the TCP transport

s0undt3ch committed:
Add TCP transport nox sessions

s0undt3ch committed:
Update compiled requirements

s0undt3ch committed:
Ignore the generated docs archive

s0undt3ch committed:
Blacklist pycrypto and install pycryptodome instead

s0undt3ch committed:
Update PR CI jobs for nox

s0undt3ch committed:
Blacklist `enum34` on Py>=3.4. Update requirements.

s0undt3ch committed:
Fix failed tests re-run logic

s0undt3ch committed:
The output is important to know why it failed.

s0undt3ch committed:
Watch out for too long shebang's

s0undt3ch committed:
Lock supervisor to 3.3.5, last kown working(passing tests) version

dwoz committed:
Create the roster even when not running ssh

dwoz committed:
Syndic roster not in 2017.7.9

waynew committed:
Update CodeClimate settings

https://codeclimate.com/github/saltstack/salt/pull/51461 had a matching
block with mass of 38. Apparently that's too low because that code was
not even the same at all. Bumping it up a bit to avoid false positives.

Also disable argument count checks because we really don't need that
here.

terminalmage committed:
Fix deprecation warning in msgpack >= 0.5.2

This prevents a deprecation warning due to the "encoding" argument being
deprecated in version 0.5.2.

s0undt3ch committed:
More insight(context) on failure reports

s0undt3ch committed:
Don't try to change ownership on non existing paths

s0undt3ch committed:
Don't fail the build if the issue was sending a slack notification

s0undt3ch committed:
These tests aren't destructive

s0undt3ch committed:
Be verbose in a CI context

s0undt3ch committed:
Strings and integers are not comparable under Python 3

s0undt3ch committed:
Try and address the fact that some python packages are only provided by the distro

This is the case for python-apt on debian based distributions.

s0undt3ch committed:
Try to make sure some pip internal functions are always present in sys.modules after reloads

s0undt3ch committed:
Actually write the lint reports.

It's just too much trouble to stream both to stdout and to a file from a
subprocess.

s0undt3ch committed:
Also ignore multiprocessing coverage files

dwoz committed:
Fix integration.client.test_kwarg test timeouts

dwoz committed:
Wait longer for ping reaction

Wait longer for the ping reaction in case there are multiple events on
the bus and/or high load on the system running tests

dwoz committed:
Do not limit event assert to first event

dwoz committed:
Fix typo

dwoz committed:
Ignore missing variable in six module

dwoz committed:
fix linter

s0undt3ch committed:
Disable re-runing failed tests for now

s0undt3ch committed:
Remove unused argument

s0undt3ch committed:
Fix docker entries

dwoz committed:
check timeout when queue empty

s0undt3ch committed:
Let's fully unload pip

s0undt3ch committed:
Specify the SHA we need

s0undt3ch committed:
Don't fail if pkg_resources is not importable

s0undt3ch committed:
Fix pip tests, in particular under windows

s0undt3ch committed:
We must also provide `venv_bin` when running within a virtualenv

s0undt3ch committed:
Add a helper to return the path to the real pytohn executable

In case the test suite is running from within a virtualenv

s0undt3ch committed:
Use the real python executable path when running within a virtualenv

s0undt3ch committed:
Disable progress bars on pip installs

s0undt3ch committed:
We don't need to be root to run these tests

s0undt3ch committed:
We need to make sure the virtualenv path entry is removed when searching

Ch3LL committed:
Fix pylint on 2017.7.9

s0undt3ch committed:
Be aware of the different path separators

s0undt3ch committed:
Default to Cryptodome, fallback to PyCrypto

s0undt3ch committed:
Skip test if required cypto libs are not available

waynew committed:
We have long files

s0undt3ch committed:
Update CI jobs to point to 2018.3

s0undt3ch committed:
Lint fixes

s0undt3ch committed:
Disable code coverage uploads.

We're not running the full test suite

s0undt3ch committed:
Update static requirements for 2018.3 branch

waynew committed:
Use functools.wraps with decorators

Otherwise, we're not fully wrapping the function so expected attributes
(at least __opts__ anyway) go missing.

Fixes #44639

waynew committed:
Add tests for wraps

During testing I also discovered that `_ignores_kwargs` was also missing
the @wraps decorator. Now it has it.

waynew committed:
Lint cleanup

s0undt3ch committed:
Add a cloud nox session for both runtests and pytest

s0undt3ch committed:
Add a nox session for the Jenkins tornado jobs

dwoz committed:
Fix timeout logic

dwoz committed:
Revert part of 927219c since it is not a fix

dwoz committed:
Avoid race condition in even assertions

dwoz committed:
Fix xml block causing docs to fail

s0undt3ch committed:
Create nox lint virtualenvs before running them

s0undt3ch committed:
Allow `--install-only` to gather the required information

s0undt3ch committed:
`impacket` does not support Py3

s0undt3ch committed:
Fix proxy minion startup issue on the test suite

dwoz committed:
Fix windows tests

dwoz committed:
Fix unit.test_loader.LoaderGlobalsTest.test_states

s0undt3ch committed:
Always combine and generate the XML coverage report

s0undt3ch committed:
Use the system's path separator

s0undt3ch committed:
Each generated script is now prepared for code coverage

s0undt3ch committed:
Include `COVERAGE_FILE` as an env variable.

This will tell coverage where to save all coverage files.

s0undt3ch committed:
As a script, not as a module

s0undt3ch committed:
More entries to ignore

s0undt3ch committed:
Limit and reduce the ammount of log records sent over the wire

s0undt3ch committed:
Upgrade etcd to > 0.4.2

dwoz committed:
Add 'runFull' build parameter for PR tests

dwoz committed:
Remove ref restriciton from kitche-salt

twangboy committed:
Skip tests that fail on Windows

twangboy committed:
Stop w32time before and start after testing

Ch3LL committed:
Update Issue Template: questions removed

Ch3LL committed:
Make sure --run-expensive runtests.py arg works

s0undt3ch committed:
Upload coverage reports on full test runs

s0undt3ch committed:
`identical_signature_wrapper` adds `__wrapped__` to function globals

dwoz committed:
Pin kubernetes to <4.0

s0undt3ch committed:
Add `moto` to the windows dependencies

s0undt3ch committed:
Install mock on Python < 3.6

dwoz committed:
Skip libcloud unit tests when no libcloud

dwoz committed:
Fix linter

dwoz committed:
Ignore super not called

dwoz committed:
Dont fail just because some random process died

dwoz committed:
Add centos-6-py2 and debain 8,9 py2 and 3 to PR tests

dwoz committed:
Skip tests when no libcloud

dwoz committed:
Fix linter

s0undt3ch committed:
Rename kitchen-centos6-py3 to kitchen-centos6-py2

s0undt3ch committed:
Fix `NOX_ENV_NAME`

s0undt3ch committed:
Fix `NOX_ENV_NAME`

s0undt3ch committed:
Fix `NOX_ENV_NAME`

dwoz committed:
Fix underscore wart (and linter)

dwoz committed:
Fix broken pip state

dwoz committed:
Fix linter

dwoz committed:
Fix unused import

dwoz committed:
Skip tests when no libcloud

dwoz committed:
fix linter

dwoz committed:
Fix missing class wart

dwoz committed:
Change py3 windows timeout from 6 to 8 hours

dwoz committed:
Fix remaining failures on ubuntu 1404 and add to PR tests

dwoz committed:
Fix test_doc on windows by using grep yay!

dwoz committed:
Fix linter

dwoz committed:
SIGKILL is not alwasy available use a local variable

dwoz committed:
Add ubuntu1804 to pr tests

dwoz committed:
Fix non raw msg pack msg decoding

dwoz committed:
Add regression test

dwoz committed:
msgpack returns string_types

dwoz committed:
Less flaky mine tests

s0undt3ch committed:
Proper english in requirements comments

s0undt3ch commented:
re-run full all

waynew commented on tests/support/gitfs.py:
```diff
@@ -86,11 +86,16 @@ def _search(proc):
-                elif name in proc.name() and _search(proc):
-                    return proc
+                except psutil.NoSuchProcess:
+                    # Whichever process we are interigating is no longer alive.
```
Should this be `interrogating`?

s0undt3ch replied:
Probably...

s0undt3ch committed:
We no longer test againt Ubuntu 14.04 which reached EOL

s0undt3ch committed:
Typo

s0undt3ch commented:
The windows failing test `unit.states.test_file.TestFileState.test_file_managed_should_fall_back_to_binary` Is being fixed in https://github.com/saltstack/salt/pull/52770

s0undt3ch committed:
Test is not destructive.

Changes salt's testing config, which is what testing is all about, but
does not change the underlying OS.

s0undt3ch commented:
This should be good to go in.
Any other failures on 2019.2.1 can be addressed in particular.

bryceml commented on requirements/static/docs.in:
```diff
@@ -0,0 +1 @@
+sphinx
```
Should we require sphinx greater than or equal to 1.7.1 since we've made modifications that require that now?

bryceml replied:
We may even need a newer version such as 1.7.2 than that to avoid warnings.
1.8.5 on python2 and 2.0.1 on python3 would be ideal.

bryceml replied:
I supposed that should be done in a separate pr though, I just saw this and thought about it.

s0undt3ch replied:
We should definitely bring in the work you've done, but yeah, lets do that on a separate PR.
